lib/matplotlib/streamplot.py

```diff
@@ -516,18 +520,30 @@ def _integrate_rk12(x0, y0, dmap, f, maxlength):
     xf_traj = []
     yf_traj = []
 
-    while dmap.grid.within_grid(xi, yi):
-        xf_traj.append(xi)
-        yf_traj.append(yi)
+    while True:
         try:
+            if dmap.grid.within_grid(xi, yi):
+                xf_traj.append(xi)
+                yf_traj.append(yi)
+            else:
+                raise OutOfBounds
+
             k1x, k1y = f(xi, yi)
-            k2x, k2y = f(xi + ds * k1x,
-                         yi + ds * k1y)
-        except IndexError:
-            # Out of the domain on one of the intermediate integration steps.
-            # Take an Euler step to the boundary to improve neatness.
-            ds, xf_traj, yf_traj = _euler_step(xf_traj, yf_traj, dmap, f)
-            stotal += ds
+
+            if dmap.grid.within_grid(xi + ds * k1x, yi + ds * k1y):
+                k2x, k2y = f(xi + ds * k1x,
+                             yi + ds * k1y)
+            else:
+                raise OutOfBounds
+
+        except OutOfBounds:
```

Comment:
If all of the raising logic is here it might be clearer to just use `if` and `else` statements.

Reply:
Another simple approach could be to move the bounds check and `raise OutOfbounds` into `f`. Would simplify the logic in the interpolator.

Reply:
I've pushed a change to the PR moving the `raise OutOfBounds` into `f`. I tried quite a few things and I think this is the clearest.